### PR TITLE
Faster room joins: Add edge case tests for outgoing device list updates

### PR DESCRIPTION
**Pull request by squahtx**
Test that outgoing device list updates reach servers incorrectly thought
to not be in a partial state room once the room transitions to full
state.

---

Built on top of #473.
Best reviewed commit by commit.

### CODE REVIEW

**Comment by squahtx** [comment]
We expect these tests to fail until https://github.com/matrix-org/synapse/issues/13891 is ready.

**Review by erikjohnston — APPROVED** [review]
The logic for some of these tests are quite confusing, but I don't have any suggestions on how to make it better :(

File: tests/federation_room_join_partial_state_test.go [hunk 1]
```diff
@@ -1719,6 +1719,246 @@ func TestPartialStateJoin(t *testing.T) {
 			mustNotReceiveDeviceListUpdate(t, deviceListUpdateChannel2, "@elsie received device list update unexpectedly.")
 			t.Log("@charlie and @derek received device list update.")
 		})
+
+		// setupIncorrectlyAcceptedKick joins the homeserver under test to a room, then joins
+		// @elsie:server2 and sends an invalid event to kick @elsie:server2 from the room.
+		setupIncorrectlyAcceptedKick := func(
```
[the remark this answers is not in the document]
Comment (squahtx) [reply tests/federation_room_join_partial_state_test.go:1727]: Good idea, let's do that.
